dev/sql_auditing.h

```diff
@@ -0,0 +1,146 @@
+#pragma once
+
+#include <fstream>
+#include <chrono>
+#include <ctime>  // std::localtime and std::tm and std::time_t
+#include <iomanip>  // std::put_time
+#include <memory>
+#include <string>
+#include <memory>
+#include "error_code.h"
+
+enum class auditing_behavior : signed char { OFF = 0, ON = 1 };
+
+namespace sqlite_orm {
+    namespace internal {
+        class storage_base;
+    }
+}
+using sqlite_orm::internal::storage_base;
+
+struct log_settings {
+    auditing_behavior behavior;
+    std::shared_ptr<std::ofstream> log_file;
+    std::string log_file_name;
+    std::string time_format;
+    enum status_code { FAIL, GOOD } status = status_code::FAIL;
+    bool global_settings = false;
+
+  public:
+    log_settings(auditing_behavior behave,
+                 std::shared_ptr<std::ofstream> file,
+                 std::string file_name,
+                 std::string format,
+                 status_code stats,
+                 bool glob_settings) {
+        behavior = behave;
+        log_file = file;
+        bool ok = log_file->good();
+        bool open = log_file->is_open();
+        log_file_name = file_name;
+        time_format = format;
+        status = stats;
+        global_settings = glob_settings;
+    }
+    log_settings() {}
+    void open() {
+        using namespace std;
+        if(!log_file->is_open()) {
+            log_file->open(log_file_name, ios::trunc | ios::out);
+            if(!log_file->good()) {
+                status = FAIL;
+                log_file->close();
+                throw std::system_error{sqlite_orm::orm_error_code::failure_to_init_logfile};
+            }
+        }
+        status = GOOD;
+    }
+    void log(std::string message) {
+        if(status == GOOD && behavior == auditing_behavior::ON) {
+            // would use format if C++ 20
+            auto now = std::chrono::system_clock::now();
+
+            std::time_t now_time = std::chrono::system_clock::to_time_t(now);
+
+            // Convert to local time (std::tm structure)
+            // WARNING: localtime is not thread safe!
+            std::tm local_time = *std::localtime(&now_time);
+
+            // Print the local time in a human-readable format
+            *log_file << "@: " << std::put_time(&local_time, time_format.c_str()) << " = ";
+            *log_file << message << std::endl;
+        }
+    }
+};
+
+inline std::shared_ptr<std::ofstream> get_file(std::string log_file_name) {
+    static std::map<std::string, std::shared_ptr<std::ofstream>> files;
+    auto it = files.find(log_file_name);
+    if(it == files.end()) {
+        std::shared_ptr<std::ofstream> pointer_to_shared{new std::ofstream()};
```

Comment:
use `std::make_shared` instead of `new`. Avoid using `new` and `delete` if you can